README.md
<p align="center"><a href="https://github.com/jef/streetmerchant#readme"><img src="https://raw.githubusercontent.com/jef/streetmerchant/main/media/streetmerchant.png" alt="streetmerchant" /></a></p>
<p align="center">The world's easiest, most powerful stock checker
<br/><br/>
<a href="https://github.com/jef/streetmerchant/actions?query=workflow%3Aci"><img src="https://github.com/jef/streetmerchant/workflows/ci/badge.svg" /></a>
<a href="https://discord.gg/gbVY4vB9JF"><img src="https://img.shields.io/discord/773913070665859073.svg?label=chat&logo=discord&logoColor=ffffff&color=7389D8" alt="Tweet" /></a>
<a href="https://twitter.com/intent/tweet?text=Beat%20the%20masses%20with%20streetmerchant&url=https://github.com/jef/streetmerchant&hashtags=typescript,opensource,bot,shopping"><img src="https://img.shields.io/badge/twitter-share-green?logo=twitter&style=social" alt="Tweet" /></a>
<br/><br/>
<a href="https://github.com/jef/streetmerchant#faq">FAQ</a> |
<a href="https://github.com/jef/streetmerchant/issues">Issues</a> |
<a href="https://github.com/jef/streetmerchant/wiki">Wiki</a>
</p>

## Features

- Scrapes multiple websites for patterns of being stocked via API and Chromium
- Opens browser when stock is available
- Ability to send notifications when stock is available

> :point_right: The bot _will not_ automatically buy for you

## Installation overview

Linux, macOS, and Windows are all capable operating systems.

You do not need any computer skills, smarts, or anything of that nature. You are very capable as you have made it this
far. Some basic understanding how a terminal, git, and or Node.js is a bonus, but that does not limit you to
getting `streetmerchant` running!

### Installation: native

| Reference | Note |
|:---:|---|
| tag | Example, `v1.0.0`; stable |
| `main` | Latest HEAD; not tagged, could be unstable |

- [Node.js 14](https://nodejs.org/en/)
- [git](https://git-scm.com/)
- Clone this project `git clone https://github.com/jef/streetmerchant.git`
  - To checkout a particular ref, use `git checkout <ref name>` after cloning
- Run `npm install`
- Copy `.env-example` to a new file `.env` and edit the `.env` file to your liking using
  your [favorite text editor](https://code.visualstudio.com/)
  - More on this in [customization](#Customization)
- Run `npm run start` to start

At any point you want the program to stop, use <kbd>Ctrl</kbd> + <kbd>C</kbd>.

> :point_right: Please visit the [wiki](https://github.com/jef/streetmerchant/wiki) if you need more help with installation.

### Installation: Docker

Available via GitHub Container Registry.

| Tag | Note |
|:---:|---|
| `latest` | Latest release; stable |
| `nightly` | Latest HEAD each day at midnight UTC; could be unstable |

```sh
# to run
docker run --cap-add=SYS_ADMIN \
  -it --rm --env-file ./.env \
  ghcr.io/jef/streetmerchant:nightly

# to test notifications
docker run --cap-add=SYS_ADMIN \
  -it --rm --env-file ./.env \
  ghcr.io/jef/streetmerchant:nightly test:notification:production
```

### Developer notes

The command `npm run start:dev` can be used instead of `npm run start` to automatically restart the project when
filesystem changes are detected in the `src/` folder or `.env` file.

## Customization

To customize `streetmerchant`, make a copy of `.env-example` as `.env` and make any changes to your liking. _All
environment variables are **optional**._

<details>
<summary>Expand to see all available options</summary>

### Application

| Environment variable | Description | Notes |
|:---:|---|---|
| `AUTO_ADD_TO_CART` | Enable auto add to cart on support stores | Default: `true` |
| `BROWSER_TRUSTED` | Skip Chromium Sandbox | Useful for containerized environments, default: `false` |
| `HEADLESS` | Puppeteer to run headless or not | Debugging related, default: `true` |
| `INCOGNITO` | Puppeteer to run incognito or not | Debugging related, default: `false` |
| `IN_STOCK_WAIT_TIME` | Time to wait between requests to the same link if it has that card in stock | In seconds, default: `0` |
| `LOG_LEVEL` | [Logging levels](https://github.com/winstonjs/winston#logging-levels) | Debugging related, default: `info` |
| `LOW_BANDWIDTH` | Blocks images/fonts to reduce traffic | Disables ad blocker, default: `false` |
| `OPEN_BROWSER` | Toggle for whether or not the browser should open when item is found | Default: `true` |
| `PAGE_BACKOFF_MIN` | Minimum backoff time between retrying requests for the same store when a forbidden response is received | Default: `10000` |
| `PAGE_BACKOFF_MAX` | Maximum backoff time between retrying requests for the same store when a forbidden response is received | Default: `3600000` |
| `PAGE_SLEEP_MIN` | Minimum sleep time between queries of the same product page | In milliseconds, default: `5000` |
| `PAGE_SLEEP_MAX` | Maximum sleep time between queries of the same product page | In milliseconds, default: `10000` |
| `PAGE_TIMEOUT` | Navigation Timeout in milliseconds | `0` for infinite, default: `30000` |
| `PROXY_PROTOCOL` | protocol of proxy server, such as `socks5` | default: `http` |
| `PROXY_ADDRESS` | IP Address or fqdn of proxy server |
| `PROXY_PORT` | TCP Port number on which the proxy is listening for connections | Default: `80` |
| `SCREENSHOT` | Capture screenshot of page if a card is found | Default: `true` |
| `USER_AGENT` | Custom User-Agents headers for HTTP requests | Newline separated, e.g.: `USER_AGENT_STRING1 \n USER_AGENT_STRING2` | | Default: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36` |
| `WEB_PORT` | Starts a webserver to be able to control the bot while it is running; optional | Default: disabled |

> :point_right: You can find your computer's user agent by [searching google for "my user agent"](http://google.com/search?q=my+user+agent)

> :point_right: Data usage is [known to be high](https://github.com/jef/streetmerchant/issues?q=is%3Aissue+sort%3Aupdated-desc+bandwidth). This is expected as the program scrapes many websites in parallel 24/7. To help reduce this, use `LOW_BANDWIDTH="true"`. We are looking into other solutions as well, but is low priority.

### Filters

| Environment variable | Description | Notes |
|:---:|---|---|
| `COUNTRY` | [Supported country](#supported-countries) you want to be scraped | Only used with `nvidia-api`, default: `usa` |
| `MAX_PRICE_SERIES_3060TI` | Maximum price allowed for a match, applies 3060 Ti series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MAX_PRICE_SERIES_3070` | Maximum price allowed for a match, applies 3070 series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MAX_PRICE_SERIES_3080` | Maximum price allowed for a match, applies 3080 series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MAX_PRICE_SERIES_3090` | Maximum price allowed for a match, applies 3090 series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MAX_PRICE_SERIES_CORSAIR_SF` | Maximum price allowed for a match, applies to Corsair PSUs | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - PSUs above `1234` will be skipped. |
| `MAX_PRICE_SERIES_RYZEN5600` | Maximum price allowed for a match, applies AMD 5600 series cpus | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - CPUs above `1234` will be skipped. |
| `MAX_PRICE_SERIES_RYZEN5800` | Maximum price allowed for a match, applies AMD 5800 series cpus | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - CPUs above `1234` will be skipped. |
| `MAX_PRICE_SERIES_RYZEN5900` | Maximum price allowed for a match, applies AMD 5900 series cpus | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - CPUs above `1234` will be skipped. |
| `MAX_PRICE_SERIES_RYZEN5950` | Maximum price allowed for a match, applies AMD 5950 series cpus | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - CPUs above `1234` will be skipped. |
| `MAX_PRICE_SERIES_SONYPS5C` | Maximum price allowed for a match, applies PS5 console | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - PS5 above `1234` will be skipped. |
| `MAX_PRICE_SERIES_SONYPS5DE` | Maximum price allowed for a match, applies PS5 digital edition | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - PS5 above `1234` will be skipped. |
| `MAX_PRICE_SERIES_TEST` | Maximum price allowed for a match, applies `test:series` | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - PS5 above `1234` will be skipped. |
| `MICROCENTER_LOCATION` | Specific MicroCenter location(s) to search | Comma separated, e.g.: `marietta,duluth`, default: `web` |
| `NVIDIA_ADD_TO_CART_ATTEMPTS` | The maximum number of times the `nvidia-api` add to cart feature will be attempted before failing | Default: `10` |
| `NVIDIA_SESSION_TTL` | The time in milliseconds to keep the cart active while using `nvidia-api` | Default: `60000` |
| `SHOW_ONLY_BRANDS` | Filter to show specified brands | Comma separated, e.g.: `evga,zotac` |
| `SHOW_ONLY_MODELS` | Filter to show specified models | Both supported formats are comma separated <br/><br/>1. Standard  E.g.: `founders edition,rog strix` <br/><br/> 2. Advanced E.g: `MODEL:SERIES`, E.g: `founders edition:3090,rog strix` |
| `SHOW_ONLY_SERIES` | Filter to show specified series | Comma separated, e.g.: `3080,ryzen5900` |

CHANGELOG.md
# Changelog

## [2.0.0](https://www.github.com/styler2go/nvidia-snatcher/compare/v1.5.0...v2.0.0) (2021-02-18)


### ⚠ BREAKING CHANGES

* removed duplicated `Model`
* update naming to `streetmerchant`

### Features

* add `meta` to logger ([#437](https://www.github.com/styler2go/nvidia-snatcher/issues/437)) ([ae6bc86](https://www.github.com/styler2go/nvidia-snatcher/commit/ae6bc86bcb75c154a2a68adda324f34f18281700))
* add `norway` to `nvidia-api` ([#304](https://www.github.com/styler2go/nvidia-snatcher/issues/304)) ([eda6c85](https://www.github.com/styler2go/nvidia-snatcher/commit/eda6c85fc03a70c5933308e96c572a480bb6c8a0))
* add alternate.nl support ([#552](https://www.github.com/styler2go/nvidia-snatcher/issues/552)) ([aca6523](https://www.github.com/styler2go/nvidia-snatcher/commit/aca65233339ddbbcc1dbd06f9011d6d44c6650cf))
* add amazon-fr to stores. ([#766](https://www.github.com/styler2go/nvidia-snatcher/issues/766)) ([02d7c47](https://www.github.com/styler2go/nvidia-snatcher/commit/02d7c47c4093ad71dfab31f93a9b1390f4cad8a0))
* add asus-de store and asus 3080 tuf/oc to amazon-de ([#436](https://www.github.com/styler2go/nvidia-snatcher/issues/436)) ([b4d8733](https://www.github.com/styler2go/nvidia-snatcher/commit/b4d8733d1823e2fa5d3a1b80d2d4218bfa737cd7))
* add browser opening to test:notification ([f87053c](https://www.github.com/styler2go/nvidia-snatcher/commit/f87053cb02e04b3cb2c0cf253187a9e9857c9858))
* add c8 and mocha for testing ([f87053c](https://www.github.com/styler2go/nvidia-snatcher/commit/f87053cb02e04b3cb2c0cf253187a9e9857c9858))
* add caseking and proshop-de store ([#521](https://www.github.com/styler2go/nvidia-snatcher/issues/521)) ([1aaa102](https://www.github.com/styler2go/nvidia-snatcher/commit/1aaa102c06281b60da593d04ced01f8b20a37a3b))
* add command parameter to set custom dotenv conf ([#1590](https://www.github.com/styler2go/nvidia-snatcher/issues/1590)) ([b03d242](https://www.github.com/styler2go/nvidia-snatcher/commit/b03d24217f8ff5a56ddf19c019d10b48afec813f))
* add coolblue store ([#482](https://www.github.com/styler2go/nvidia-snatcher/issues/482)) ([5c61333](https://www.github.com/styler2go/nvidia-snatcher/commit/5c61333df00c0c30f04e4aced01a498c5816ad58))
* add deprecation notices ([601742c](https://www.github.com/styler2go/nvidia-snatcher/commit/601742cf2c0b5ccf0131e4c4e904ad4037d80a8d))
* add german stores ([#462](https://www.github.com/styler2go/nvidia-snatcher/issues/462)) ([85a07dc](https://www.github.com/styler2go/nvidia-snatcher/commit/85a07dc0dc6a8dfe2aba1d4ea3e4cde6ec083086))
* add incognito mode ([#534](https://www.github.com/styler2go/nvidia-snatcher/issues/534)) ([2b6457e](https://www.github.com/styler2go/nvidia-snatcher/commit/2b6457e8409658b843204fa0aeafb00d48f9e53e))
* add link series to error message for better context ([#265](https://www.github.com/styler2go/nvidia-snatcher/issues/265)) ([332b4a8](https://www.github.com/styler2go/nvidia-snatcher/commit/332b4a8246320e458729c0b58c310d290df12530))
* add option to disable auto add to cart ([#759](https://www.github.com/styler2go/nvidia-snatcher/issues/759)) ([a5249ec](https://www.github.com/styler2go/nvidia-snatcher/commit/a5249ecc7c3acc29a1f9d5285bd77397263cb135))
* Add PagerDuty Integration ([#565](https://www.github.com/styler2go/nvidia-snatcher/issues/565)) ([11ee0bf](https://www.github.com/styler2go/nvidia-snatcher/commit/11ee0bf1a3e2d0d550d2731f9de9e178c382ea1f))
* add production scripts and better `Dockerfile` ([0d14500](https://www.github.com/styler2go/nvidia-snatcher/commit/0d14500d5d8b93204bb99baf34dc9de745ae645d)), closes [#710](https://www.github.com/styler2go/nvidia-snatcher/issues/710)
* add proshop-dk ([#524](https://www.github.com/styler2go/nvidia-snatcher/issues/524)) ([dc2fcf5](https://www.github.com/styler2go/nvidia-snatcher/commit/dc2fcf5ab4dba8be459bd5aa6a41e3e097205d45))
* add proxy rotation (settable per store) ([#1026](https://www.github.com/styler2go/nvidia-snatcher/issues/1026)) ([490d44e](https://www.github.com/styler2go/nvidia-snatcher/commit/490d44e1fdfdf91db7344dd3050ba8fd8d069631))
* add ps5 max price ([30a51a3](https://www.github.com/styler2go/nvidia-snatcher/commit/30a51a38bee1c399276466e5db457bbe702edd5c))
* add redis ([#1390](https://www.github.com/styler2go/nvidia-snatcher/issues/1390)) ([fb82526](https://www.github.com/styler2go/nvidia-snatcher/commit/fb82526a427e503fff1f1cf56b091ff0c8a5a96e))
* add restart time for leaking Chromium in Ubuntu ([#1880](https://www.github.com/styler2go/nvidia-snatcher/issues/1880)) ([1009ce6](https://www.github.com/styler2go/nvidia-snatcher/commit/1009ce6636c21c378121bd83b251a371a1b8568e))
* add rog strix oc to amazon-de ([#471](https://www.github.com/styler2go/nvidia-snatcher/issues/471)) ([e5909c5](https://www.github.com/styler2go/nvidia-snatcher/commit/e5909c532da1c8b731e9835bffd8f31b76550d4d))
* add series to Discord webhook ([#618](https://www.github.com/styler2go/nvidia-snatcher/issues/618)) ([273d058](https://www.github.com/styler2go/nvidia-snatcher/commit/273d05841451ed896f1d505bcd24f2230a143d3a))
* add support for multiple browser user agents ([#547](https://www.github.com/styler2go/nvidia-snatcher/issues/547)) ([10a81dc](https://www.github.com/styler2go/nvidia-snatcher/commit/10a81dcbd54557652537cb641cbbd727d39e34e5))
* add support for specifying smtp server ([#458](https://www.github.com/styler2go/nvidia-snatcher/issues/458)) ([160ae37](https://www.github.com/styler2go/nvidia-snatcher/commit/160ae37d7bee4f0323cb939566d586f1300aece6))

src/config.ts
import {banner} from './banner';

import {config as config_} from 'dotenv';
import path from 'path';
import {readFileSync} from 'fs';

config_({path: path.resolve(__dirname, '../.env')});

console.info(
	banner.render(
		envOrBoolean(process.env.ASCII_BANNER, false),
		envOrString(process.env.BANNER_COLOR, '#808080')
	)
);

/**
 * Returns environment variable, given array, or default array.
 *
 * @param environment Interested environment variable.
 * @param array Default array. If not set, is `[]`.
 */
function envOrArray(
	environment: string | undefined,
	array?: string[]
): string[] {
	return (environment
		? environment.includes('\n')
			? environment.split('\n')
			: environment.split(',')
		: array ?? []
	).map((s) => s.trim());
}

/**
 * Returns environment variable, given boolean, or default boolean.
 *
 * @param environment Interested environment variable.
 * @param boolean Default boolean. If not set, is `true`.
 */
function envOrBoolean(
	environment: string | undefined,
	boolean?: boolean
): boolean {
	return environment ? environment === 'true' : boolean ?? true;
}

/**
 * Returns environment variable, given string, or default string.
 *
 * @param environment Interested environment variable.
 * @param string Default string. If not set, is `''`.
 */
function envOrString(environment: string | undefined, string?: string): string {
	return environment ? environment : string ?? '';
}

/**
 * Returns environment variable, given number, or default number.
 *
 * @param environment Interested environment variable.
 * @param number Default number. If not set, is `0`.
 */
function envOrNumber(environment: string | undefined, number?: number): number {
	return environment ? Number(environment) : number ?? 0;
}

/**
 * Returns environment variable, given number, or default number,
 * while handling .env input errors for a Min/Max pair.
 * .env errors handled:
 * - Min/Max swapped (Min larger than Max, Max smaller than Min)
 * - Min larger than default Max when no Max defined
 * - Max smaller than default Min when no Min defined
 *
 * @param environmentMin Min environment variable of Min/Max pair.
 * @param environmentMax Max environment variable of Min/Max pair.
 * @param number Default number. If not set, is `0`.
 */
function envOrNumberMin(
	environmentMin: string | undefined,
	environmentMax: string | undefined,
	number?: number
) {
	if (environmentMin || environmentMax) {
		if (environmentMin && environmentMax) {
			return Number(
				Number(environmentMin) < Number(environmentMax)
					? environmentMin
					: environmentMax
			);
		}

		if (environmentMax) {
			return Number(environmentMax) < (number ?? 0)
				? Number(environmentMax)
				: number ?? 0;
		}

		if (environmentMin) {
			return Number(environmentMin);
		}
	}

	return number ?? 0;
}

/**
 * Returns environment variable, given number, or default number,
 * while handling .env input errors for a Min/Max pair.
 * .env errors handled:
 * - Min/Max swapped (Min larger than Max, Max smaller than Min)
 * - Min larger than default Max when no Max defined
 * - Max smaller than default Min when no Min defined
 *
 * @param environmentMin Min environment variable of Min/Max pair.
 * @param environmentMax Max environment variable of Min/Max pair.
 * @param number Default number. If not set, is `0`.
 */
function envOrNumberMax(
	environmentMin: string | undefined,
	environmentMax: string | undefined,
	number?: number
) {
	if (environmentMin || environmentMax) {
		if (environmentMin && environmentMax) {
			return Number(
				Number(environmentMin) < Number(environmentMax)
					? environmentMax
					: environmentMax
			);
		}

		if (environmentMin) {
			return Number(environmentMin) > (number ?? 0)
				? Number(environmentMin)
				: number ?? 0;
		}

		if (environmentMax) {
			return Number(environmentMax);
		}
	}

	return number ?? 0;
}

const browser = {
	isHeadless: envOrBoolean(process.env.HEADLESS),
	isIncognito: envOrBoolean(process.env.INCOGNITO, false),
	isTrusted: envOrBoolean(process.env.BROWSER_TRUSTED, false),
	lowBandwidth: envOrBoolean(process.env.LOW_BANDWIDTH, false),
	maxBackoff: envOrNumberMax(
		process.env.PAGE_BACKOFF_MIN,
		process.env.PAGE_BACKOFF_MAX,
		3600000
	),
	maxSleep: envOrNumberMax(
		process.env.PAGE_SLEEP_MIN,
		process.env.PAGE_SLEEP_MAX,
		10000
	),
	minBackoff: envOrNumberMin(
		process.env.PAGE_BACKOFF_MIN,
		process.env.PAGE_BACKOFF_MAX,
		10000
	),
	minSleep: envOrNumberMin(
		process.env.PAGE_SLEEP_MIN,
		process.env.PAGE_SLEEP_MAX,
		5000
	),
	open: envOrBoolean(process.env.OPEN_BROWSER)
};

const docker = envOrBoolean(process.env.DOCKER);

const logLevel = envOrString(process.env.LOG_LEVEL, 'info');

const notifications = {
	desktop: process.env.DESKTOP_NOTIFICATIONS === 'true',
	discord: {
		notifyGroupSeries: {
			3070: envOrString(process.env.DISCORD_NOTIFY_GROUP_3070),
			3080: envOrString(process.env.DISCORD_NOTIFY_GROUP_3080),
			3090: envOrString(process.env.DISCORD_NOTIFY_GROUP_3090),
			rx6800: envOrString(process.env.DISCORD_NOTIFY_GROUP_RX6800),
			rx6800xt: envOrString(process.env.DISCORD_NOTIFY_GROUP_RX6800XT),
			rx6900xt: envOrString(process.env.DISCORD_NOTIFY_GROUP_RX6900XT),
			ryzen5600: envOrString(process.env.DISCORD_NOTIFY_GROUP_RYZEN5600),
			ryzen5800: envOrString(process.env.DISCORD_NOTIFY_GROUP_RYZEN5800),
			ryzen5900: envOrString(process.env.DISCORD_NOTIFY_GROUP_RYZEN5900),
			ryzen5950: envOrString(process.env.DISCORD_NOTIFY_GROUP_RYZEN5950)
		},
		webHookUrl: envOrArray(process.env.DISCORD_WEB_HOOK)
	},
	email: {
		password: envOrString(process.env.EMAIL_PASSWORD),
		smtpAddress: envOrString(process.env.SMTP_ADDRESS),
		smtpPort: envOrNumber(process.env.SMTP_PORT, 25),
		to: envOrString(
			process.env.EMAIL_TO,
			envOrString(process.env.EMAIL_USERNAME)
		),
		username: envOrString(process.env.EMAIL_USERNAME)
	},
	mqtt: {
		broker: envOrString(process.env.MQTT_BROKER_ADDRESS),
		clientId: envOrString(process.env.MQTT_CLIENT_ID),
		password: envOrString(process.env.MQTT_PASSWORD),
		port: envOrNumber(process.env.MQTT_BROKER_PORT, 1883),
		qos: envOrNumber(process.env.MQTT_QOS, 0),
		topic: envOrString(process.env.MQTT_TOPIC, 'streetmerchant/alert'),
		username: envOrString(process.env.MQTT_USERNAME)
	},
	pagerduty: {
		integrationKey: envOrString(process.env.PAGERDUTY_INTEGRATION_KEY),
		severity: envOrString(process.env.PAGERDUTY_SEVERITY, 'info')
	},
	philips_hue: {
		accessToken: envOrString(process.env.PHILIPS_HUE_CLOUD_ACCESS_TOKEN),
		apiKey: envOrString(process.env.PHILIPS_HUE_API_KEY),
		bridgeIp: envOrString(process.env.PHILIPS_HUE_LAN_BRIDGE_IP),
		clientId: envOrString(process.env.PHILIPS_HUE_CLOUD_CLIENT_ID),
		clientSecret: envOrString(process.env.PHILIPS_HUE_CLOUD_CLIENT_SECRET),
		lightColor: envOrString(process.env.PHILIPS_HUE_LIGHT_COLOR),
		lightIds: envOrString(process.env.PHILIPS_HUE_LIGHT_IDS),
		lightPattern: envOrString(process.env.PHILIPS_HUE_LIGHT_PATTERN),
		refreshToken: envOrString(process.env.PHILIPS_HUE_CLOUD_REFRESH_TOKEN),
		remoteApiUsername: envOrString(process.env.PHILIPS_HUE_API_KEY)
	},
	phone: {
		availableCarriers: new Map([
			['att', 'txt.att.net'],
			['attgo', 'mms.att.net'],
			['bell', 'txt.bell.ca'],
			['fido', 'fido.ca'],
			['google', 'msg.fi.google.com'],
			['koodo', 'msg.koodomobile.com'],
			['mint', 'mailmymobile.net'],
			['rogers', 'pcs.rogers.com'],
			['sprint', 'messaging.sprintpcs.com'],
			['telus', 'msg.telus.com'],
			['tmobile', 'tmomail.net'],
			['verizon', 'vtext.com'],
			['virgin', 'vmobl.com'],
			['virgin-ca', 'vmobile.ca']
		]),
		carrier: envOrArray(process.env.PHONE_CARRIER),
		number: envOrArray(process.env.PHONE_NUMBER)
	},
	playSound: envOrString(process.env.PLAY_SOUND),
	pushbullet: envOrString(process.env.PUSHBULLET),
	pushover: {
		priority: envOrNumber(process.env.PUSHOVER_PRIORITY),
		token: envOrString(process.env.PUSHOVER_TOKEN),
		username: envOrString(process.env.PUSHOVER_USER)
	},
	slack: {
		channel: envOrString(process.env.SLACK_CHANNEL),
		token: envOrString(process.env.SLACK_TOKEN)
	},
	telegram: {
		accessToken: envOrString(process.env.TELEGRAM_ACCESS_TOKEN),
		chatId: envOrArray(process.env.TELEGRAM_CHAT_ID)
	},
	twilio: {
		accountSid: envOrString(process.env.TWILIO_ACCOUNT_SID),
		authToken: envOrString(process.env.TWILIO_AUTH_TOKEN),
		from: envOrString(process.env.TWILIO_FROM_NUMBER),
		to: envOrString(process.env.TWILIO_TO_NUMBER)
	},
	twitch: {
		accessToken: envOrString(process.env.TWITCH_ACCESS_TOKEN),
		channel: envOrString(process.env.TWITCH_CHANNEL),
		clientId: envOrString(process.env.TWITCH_CLIENT_ID),
		clientSecret: envOrString(process.env.TWITCH_CLIENT_SECRET),
		refreshToken: envOrString(process.env.TWITCH_REFRESH_TOKEN)
	},
	twitter: {
		accessTokenKey: envOrString(process.env.TWITTER_ACCESS_TOKEN_KEY),
		accessTokenSecret: envOrString(process.env.TWITTER_ACCESS_TOKEN_SECRET),
		consumerKey: envOrString(process.env.TWITTER_CONSUMER_KEY),
		consumerSecret: envOrString(process.env.TWITTER_CONSUMER_SECRET),
		tweetTags: envOrString(process.env.TWITTER_TWEET_TAGS)
	}
};

const nvidia = {
	addToCardAttempts: envOrNumber(process.env.NVIDIA_ADD_TO_CART_ATTEMPTS, 10),
	sessionTtl: envOrNumber(process.env.NVIDIA_SESSION_TTL, 60000)
};

const page = {
	height: 1080,
	inStockWaitTime: envOrNumber(process.env.IN_STOCK_WAIT_TIME),
	screenshot: envOrBoolean(process.env.SCREENSHOT),
	timeout: envOrNumber(process.env.PAGE_TIMEOUT, 30000),
	userAgents: envOrArray(process.env.USER_AGENT, [
		'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36'
	]),
	width: 1920
};

const proxy = {
	address: envOrString(process.env.PROXY_ADDRESS),
	port: envOrNumber(process.env.PROXY_PORT, 80),
	protocol: envOrString(process.env.PROXY_PROTOCOL, 'http')
};

// Check for deprecated configuration values
if (process.env.MAX_PRICE) {
	console.warn(
		'ℹ MAX_PRICE is deprecated, please use MAX_PRICE_SERIES_{{series}}'
	);
}

const store = {
	autoAddToCart: envOrBoolean(process.env.AUTO_ADD_TO_CART, true),
	country: envOrString(process.env.COUNTRY, 'usa'),
	maxPrice: {
		series: {
			'3060ti': envOrNumber(process.env.MAX_PRICE_SERIES_3060TI),
			3070: envOrNumber(process.env.MAX_PRICE_SERIES_3070),
			3080: envOrNumber(process.env.MAX_PRICE_SERIES_3080),
			3090: envOrNumber(process.env.MAX_PRICE_SERIES_3090),
			rx6800: envOrNumber(process.env.MAX_PRICE_SERIES_RX6800),
			rx6800xt: envOrNumber(process.env.MAX_PRICE_SERIES_RX6800XT),
			rx6900xt: envOrNumber(process.env.MAX_PRICE_SERIES_RX6900XT),
			ryzen5600: envOrNumber(process.env.MAX_PRICE_SERIES_RYZEN5600),
			ryzen5800: envOrNumber(process.env.MAX_PRICE_SERIES_RYZEN5800),
			ryzen5900: envOrNumber(process.env.MAX_PRICE_SERIES_RYZEN5900),
			ryzen5950: envOrNumber(process.env.MAX_PRICE_SERIES_RYZEN5950),
			sf: envOrNumber(process.env.MAX_PRICE_SERIES_CORSAIR_SF),
			sonyps5c: envOrNumber(process.env.MAX_PRICE_SERIES_SONYPS5C),
			sonyps5de: envOrNumber(process.env.MAX_PRICE_SERIES_SONYPS5DE),
			'test:series': envOrNumber(process.env.MAX_PRICE_SERIES_TEST),
			xboxss: -1,
			xboxsx: -1
		}
	},
	microCenterLocation: envOrArray(process.env.MICROCENTER_LOCATION, ['web']),
	showOnlyBrands: envOrArray(process.env.SHOW_ONLY_BRANDS),
	showOnlyModels: envOrArray(process.env.SHOW_ONLY_MODELS).map((entry) => {
		const [name, series] = entry.match(/[^:]+/g) ?? [];
		return {
			name: envOrString(name),
			series: envOrString(series)
		};
	}),
	showOnlySeries: envOrArray(process.env.SHOW_ONLY_SERIES, [
		'3060ti',
		'3070',
		'3080',
		'3090',
		'rx6800',
		'rx6800xt',
		'rx6900xt',
		'ryzen5600',
		'ryzen5800',
		'ryzen5900',
		'ryzen5950',
		'sonyps5c',
		'sonyps5de',
		'xboxss',
		'xboxsx'
	]),
	stores: envOrArray(process.env.STORES, ['nvidia']).map((entry) => {
		const [name, minPageSleep, maxPageSleep] = entry.match(/[^:]+/g) ?? [];

		let proxyList;
		try {
			proxyList = readFileSync(`${name}.proxies`)
				.toString()
				.trim()
				.split('\n')
				.map((x) => x.trim());
		} catch {}

		return {
			maxPageSleep: envOrNumberMax(
				minPageSleep,
				maxPageSleep,
				browser.maxSleep
			),
			minPageSleep: envOrNumberMin(
				minPageSleep,
				maxPageSleep,
				browser.minSleep
			),
			name: envOrString(name),
			proxyList
		};
	})
};

export const defaultStoreData = {
	maxPageSleep: browser.maxSleep,
	minPageSleep: browser.minSleep
};

export const config = {
	browser,
	docker,
	logLevel,
	notifications,
	nvidia,
	page,
	proxy,
	store
};

export function setConfig(newConfig: any) {
	const writeConfig = config as any;
	for (const key of Object.keys(newConfig)) {
		writeConfig[key] = newConfig[key];
	}
}
